booster/library/Booster/SMT/Interface.hs

```diff
@@ -183,50 +207,86 @@ getModelFor ctxt ps subst
-                        smtTranslateError . pack $
-                            "Solver returned non-var in translation state: " <> show other
-                    sortsToTranslate = Set.fromList [SortInt, SortBool]
-
-                    (freeVarsToSExprs, untranslatableVars) =
-                        Map.partitionWithKey
-                            (const . ((`Set.member` sortsToTranslate) . (.variableSort)))
-                            freeVarsMap
-                unless (Map.null untranslatableVars) $
-                    let vars = Pretty.renderText . hsep . map pretty $ Map.keys untranslatableVars
-                     in Log.logMessage ("Untranslatable variables in model: " <> vars)
-
-                response <-
-                    if Map.null freeVarsMap
-                        then pure $ Values []
-                        else runCmd $ GetValue (Map.elems freeVarsMap)
-                runCmd_ SMT.Pop
-                case response of
-                    Error msg ->
-                        throwSMT' $ BS.unpack msg
-                    Values pairs ->
-                        let (errors, values) =
-                                Map.partition isLeft
-                                    . Map.map (valueToTerm transState)
-                                    $ Map.compose (Map.fromList pairs) freeVarsToSExprs
-                            untranslated =
-                                Map.mapWithKey (const . Var) untranslatableVars
-                         in if null errors
-                                then pure $ Right $ Map.map fromRight' values <> untranslated
-                                else
-                                    throwSMT . Text.unlines $
-                                        ( "SMT errors while converting results: "
-                                            : map fromLeft' (Map.elems errors)
-                                        )
-                    other ->
-                        throwSMT' $ "Unexpected SMT response to GetValue: " <> show other
+            Sat -> Right <$> extractModel transState
+
+    getReasonUnknown :: SMT io (Either Response (Map Variable Term))
+    getReasonUnknown = do
+        res <- runCmd SMT.GetReasonUnknown
+        runCmd_ SMT.Pop
+        pure . Left $ res
+
+    translated :: Either Text ([DeclareCommand], TranslationState)
+    translated =
+        SMT.runTranslator $ do
+            let mkSMTEquation v t =
+                    SMT.eq <$> SMT.translateTerm (Var v) <*> SMT.translateTerm t
+            smtSubst <-
+                mapM (\(v, t) -> Assert "Substitution" <$> mkSMTEquation v t) $ Map.assocs subst
+            smtPs <-
+                mapM (\(Predicate p) -> Assert (mkComment p) <$> SMT.translateTerm p) ps
+            pure $ smtSubst <> smtPs
+
+    interactWithSolver :: [DeclareCommand] -> SMT io Response
+    interactWithSolver smtAsserts = do
+        runCmd_ SMT.Push -- assuming the prelude has been run already,
```

Comment:
the pushing/popping of contexts is now a bit difficult to follow. would it not be better to push explicitly before calling `interactWithSolver` or even possibly use something like a `withSmtContext` which wraps a bunch of smt commands in a push pop block...

Reply:
Yes, the assertion stack is indeed a bit of a mess, sorry about that. I think it would be good to keep the assrtion stack manipulation confined to the `interactWithSolver` functions. I will try to clean it up.